File: src/design3/decoder/ppp_common/ppp_common/crypto_wrappers.py
"""
@file crypto_wrappers.py
@brief Crypto wrappers over Monocypher
@author Plaid Parliament of Pwning
@copyright Copyright (c) 2025 Carnegie Mellon University
"""

from .cstruct import cstruct

import monocypher
import struct

SYMMETRIC_KEY_LEN = 32
SYMMETRIC_NONCE_LEN = 24
SYMMETRIC_MAC_LEN = 16
SYMMETRIC_METADATA_LEN = SYMMETRIC_NONCE_LEN + SYMMETRIC_MAC_LEN

TREE_KEY_LEN = 16
TREE_LEFT_RIGHT_LEN = 32

PUBLIC_KEY_LEN = 64
PRIVATE_KEY_LEN = 64
SIGNATURE_LEN = 64


# Ciphertext will be length+SYMMETRIC_METADATA_LEN bytes long
# Provides authenticated encryption (any tampering will be detected upon decrypt)
# Tend to match the crypto_wrapper.c
def encrypt_symmetric(plaintext: bytes, sym_key: bytes) -> bytes:
    assert len(sym_key) == SYMMETRIC_KEY_LEN

    nonce = monocypher.generate_key(SYMMETRIC_NONCE_LEN)
    mac, ct = monocypher.lock(sym_key, nonce, plaintext)
    ciphertext = mac + nonce + ct

    assert len(ciphertext) == len(plaintext) + SYMMETRIC_METADATA_LEN
    return ciphertext


def sign_asymmetric(message: bytes, secret_key: bytes) -> bytes:
    assert len(secret_key) == PRIVATE_KEY_LEN

    out = monocypher.signature_sign(secret_key, message)

    assert len(out) == SIGNATURE_LEN
    return out


def hash_length(message: bytes, hash_size: int = 64) -> bytes:
    state = monocypher.Blake2b(hash_size=hash_size)
    state.update(message)
    hash = state.finalize()

    assert len(hash) == hash_size
    return hash


# match: crypto_wrappers.c -> kdf_tree_child() -> tmp
class TreeChildTmp(metaclass=cstruct):
    parent: bytes = f"{TREE_KEY_LEN}s"
    left_right: bytes = f"{TREE_LEFT_RIGHT_LEN}s"


def kdf_tree_child(parent_key: bytes, left_right: bytes) -> bytes:
    assert len(parent_key) == TREE_KEY_LEN
    assert len(left_right) == TREE_LEFT_RIGHT_LEN

    packed = TreeChildTmp(parent_key, left_right).pack()

    child_key = hash_length(packed, TREE_KEY_LEN)

    assert len(child_key) == TREE_KEY_LEN
    return child_key


def kdf_tree_leaf(leaf_key: bytes) -> bytes:
    assert len(leaf_key) == TREE_KEY_LEN

    frame_key = hash_length(leaf_key, SYMMETRIC_KEY_LEN)

    assert len(frame_key) == SYMMETRIC_KEY_LEN
    return frame_key


def kdf_id(id_root_key: bytes, decoder_id: int) -> bytes:
    assert len(id_root_key) == SYMMETRIC_KEY_LEN
    assert 0 <= decoder_id < 2**32

    packed = struct.pack("<I32s", decoder_id, id_root_key)

    key = hash_length(packed, SYMMETRIC_KEY_LEN)

    assert len(key) == SYMMETRIC_KEY_LEN
    return key


def kdf_symbol_shimmy(shimmy_root_key: bytes, decoder_id: int) -> bytes:
    # This happens to use the same operation so we can avoid repeating
    return kdf_id(shimmy_root_key, decoder_id)

File: src/design3/decoder/src/crypto_wrappers.c
/**
 * @file crypto_wrappers.c
 * @brief Crypto wrappers over Monocypher
 * @author Plaid Parliament of Pwning
 * @copyright Copyright (c) 2025 Carnegie Mellon University
 */

#include "crypto_wrappers.h"

#include "common.h"
#include "fiproc.h"
#include "util.h"

#include <monocypher.h>
#include <stddef.h>
#include <stdint.h>
#include <string.h>

/**
 * @brief Wrapper for symmetric decryption
 *
 * Plaintext will be length bytes long
 * Ciphertext will be length+CC_ENC_SYM_METADATA_LEN bytes long
 * Provides authenticated encryption;
 *
 * @param plaintext pointer to plaintext
 * @param ciphertext pointer to ciphertext
 * @param length length of the plaintext
 * @param sym_key symmetric key
 * @return OK if decrypt succeeds, ERROR if tampering or corruption detected
 */
error_t decrypt_symmetric(uint8_t* plaintext, const uint8_t* ciphertext, size_t length,
                          const uint8_t* sym_key) {

    volatile int res1 = crypto_aead_unlock(plaintext, ciphertext + 0, sym_key, ciphertext + 16,
                                           NULL, 0, ciphertext + 40, length);
    fiproc_delay();

    if (res1 == 0) {
        return OK;
    } else {
        return ERROR;
    }
}

/**
 * @brief Wrapper for asymmetric signature checking
 *
 * @param signature pointer to signature to verify (len 64)
 * @param message pointer to message
 * @param length length of message
 * @param pubkey pointer to public key (len 32)
 * @return OK if the signature is valid, ERROR if the signature is invalid
 */
error_t verify_asymmetric(const uint8_t* signature, const uint8_t* message, size_t length,
                          const uint8_t* pubkey) {
    volatile int res1 = crypto_eddsa_check(signature, pubkey, message, length);
    fiproc_delay();

    if (res1 == 0) {
        return OK;
    } else {
        return ERROR;
    }
}

/**
 * @brief Derives a child tree key
 *
 * @param key_out: output key (16 bytes)
 * @param parent: tree key of the parent node (16 bytes)
 * @param left_right: left or right tree key (32 bytes)
 */
void kdf_tree_child(uint8_t* key_out, const uint8_t* parent, const uint8_t* left_right) {
    UTIL_ASSERT(key_out != NULL);
    UTIL_ASSERT(parent != NULL);
    UTIL_ASSERT(left_right != NULL);

    // match: crypto_wrappers.py -> TreeChildTmp
    struct {
        uint8_t parent[TREE_KEY_LEN];
        uint8_t left_right[TREE_LEFT_RIGHT_LEN];
    } tmp;
    memcpy(tmp.parent, parent, sizeof(tmp.parent));
    memcpy(tmp.left_right, left_right, sizeof(tmp.left_right));

    // no need to do this multiple times as FI would just result in a garbage key that will fail
    // decryption
    crypto_blake2b(key_out, TREE_KEY_LEN, (uint8_t*)&tmp, sizeof(tmp));
    fiproc_delay();
}

/**
 * @brief Derives a symmetric key from a leaf tree key
 *
 * @param key_out: output key (32 bytes)
 * @param tree_key: tree key of a leaf node (16 bytes)
 */
void kdf_tree_leaf(uint8_t* key_out, const uint8_t* tree_key) {
    UTIL_ASSERT(key_out != NULL);
    UTIL_ASSERT(tree_key != NULL);

    // no need to do this multiple times as FI would just result in a garbage key that will fail
    // decryption
    crypto_blake2b(key_out, SYMMETRIC_KEY_LEN, tree_key, TREE_KEY_LEN);
    fiproc_delay();
}
